appinterface/projection/crossfire/crossfire.go

```diff
@@ -0,0 +1,754 @@
+		return fmt.Errorf("error projecting validator view: %v", err)
+	}
+
+	// Ranks computation at the end of event projection
+	if blockTime.Before(projection.competitionEndTime) {
+		if err := projection.computeTxSentRank(crossfireValidatorsStatsView, crossfireValidatorsView); err != nil {
+			return fmt.Errorf("error Updating TxSentTask Rank %v", err)
+		}
+	}
+	if blockTime.Before(projection.phaseThreeStartTime) {
+		if err := projection.computeCommitmentRank(
+			"rank_task_phase_1_2_commitment_count",
+			constants.PHASE_1N2_COMMIT_PREFIX,
+			crossfireValidatorsStatsView,
+			crossfireValidatorsView,
+		); err != nil {
+			return fmt.Errorf("error compute rank_task_phase_1_2_commitment_count %v", err)
+		}
+	}
+	if blockTime.After(projection.phaseThreeStartTime) && blockTime.Before(projection.competitionEndTime) {
+		if err := projection.computeCommitmentRank(
+			"rank_task_phase_3_commitment_count",
+			constants.PHASE_3_COMMIT_PREFIX,
+			crossfireValidatorsStatsView,
+			crossfireValidatorsView,
+		); err != nil {
+			return fmt.Errorf("error compute rank_task_phase_3_commitment_count %v", err)
+		}
+	}
+
+	// Done current height projection and ranking computation
+	if err := projection.UpdateLastHandledEventHeight(rdbTxHandle, height); err != nil {
+		return fmt.Errorf("error updating last handled event height: %v", err)
+	}
+
+	if err := rdbTx.Commit(); err != nil {
+		return fmt.Errorf("error committing changes: %v", err)
+	}
+	committed = true
+	return nil
+}
+
+func (projection *Crossfire) handleBlockCreatedEvent(
+	crossfireChainStatsView *view.CrossfireChainStats,
+	crossfireValidatorsView *view.CrossfireValidators,
+	crossfireValidatorsStatsView *view.CrossfireValidatorsStats,
+	event *event_usecase.BlockCreated,
+) error {
+	blockTime := event.Block.Time
+	blockHeight := event.Block.Height
+
+	// increment current phase block number
+	if blockTime.After(projection.phaseOneStartTime) && blockTime.Before(projection.phaseTwoStartTime) {
+		err := crossfireChainStatsView.IncrementOne(constants.PHASE_1_BLOCK_COUNT)
+		if err != nil {
+			return fmt.Errorf("error increment phase1 block count")
+		}
+	} else if blockTime.After(projection.phaseTwoStartTime) && blockTime.Before(projection.phaseThreeStartTime) {
+		err := crossfireChainStatsView.IncrementOne(constants.PHASE_2_BLOCK_COUNT)
+		if err != nil {
+			return fmt.Errorf("error increment phase2 block count")
+		}
+	} else if blockTime.After(projection.phaseThreeStartTime) && blockTime.Before(projection.competitionEndTime) {
+		// check the keep active task, throttling with every 10 blocks
+		if blockHeight%10 == 0 {
+			if err := projection.checkTaskKeepActive(crossfireChainStatsView, crossfireValidatorsView, crossfireValidatorsStatsView); err != nil {
+				return fmt.Errorf("error checkTaskKeepActive: %v", err)
+			}
+		}
```

Comment:
Why do we check phase 2 activeness only during phase 3?

Reply:
I think the active task is only part of phase 2.

Reply:
Does it mean the rank will update only after phase 2 ends?

Reply:
I am thinking if we do this check during phase 2 as well, then we will have real time (per 10 blocks) stats

Reply:
> Does it mean the rank will update only after phase 2 ends?

Yes correct.

> I am thinking if we do this check during phase 2 as well, then we will have real time (per 10 blocks) stats


Hmm. This sounds correct too.